src/main.rs

```diff
@@ -120,6 +132,66 @@ fn darker(a: u8, b: u8) -> u8 {
     a.min(b)
 }
 
+fn histogram(img: &GrayImage) -> [u32; 256] {
+    let mut result = [0; 256];
+    for p in img.pixels() {
+        result[p[0] as usize] += 1;
+    }
+    for i in 0..256 {
+        result[i] = ((result[i] as f32) + 1f32).log(1.1) as u32
+    }
+    result
+}
+
+fn smooth(histogram: &mut [u32; 256], right_to_left: bool) -> Vec<u8> {
+    let mut prev_direction = 0;
+    let mut turns = vec![];
+    for i in 0..255 {
+        let (prev_index, index, next_index) = if !right_to_left {
+            (i - 1, i, i + 1)
+        } else {
+            (256 - i, 255 - i, 254 - i)
+        };
+        histogram[index] = (histogram[index] + histogram[next_index]) / 2;
+        if i == 0 {
+            continue;
+        };
+        let direction = i64::signum(histogram[index] as i64 - histogram[prev_index] as i64);
+        if direction != 0 && direction != prev_direction {
+            if prev_direction != 0 {
+                turns.push(index as u8);
+            }
+            prev_direction = direction
+        }
+    }
+    if right_to_left {
+        turns.reverse()
+    }
+    turns
+}
+
+fn get_turns(img: &GrayImage, debug_mode: bool) -> (u8, u8) {
+    let mut h = histogram(&img);
+    let mut round = 0;
+    let turns = loop {
+        let turns = smooth(&mut h, round % 2 == 1);
+        if debug_mode {
+            println!("Turns after smoothing: {:?}", turns)
+        }
+        round += 1;
+        if turns.len() <= 3 || round > 100 {
+            break turns;
+        }
+    };
+    if turns.len() < 2 {
+        (0, 255)
+    } else {
+        let dark = 2 * turns[0];
+        let light = 255 - 2 * (255 - turns[turns.len() - 1]);
```

Comment:
`turns[turns.len() - 1] == turns.last().unwrap()`
Slightly longer but IMO reads better

Reply:
Also the element extracting part can be rewritten as a single match
```
    match turns[..] {
        [first, .., last] => (first, last),
        _ => (0, 255)
    }
```
But I am not sure how to elegantly build this into this algorithm

Reply:
Nice, I used the latter. The `[..]` matching was new to me, very elegant!